compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.td

```diff
@@ -202,5 +312,152 @@ def IREEGPU_MmaScheduleAttr : AttrDef<IREEGPU_Dialect, "MMASchedule"> {
+  let description = [{
+    This attribute contains hardware features/limits at a single GPU chip level.
+    Here a GPU chip means the hardware functionality scope where the whole
+    software compute grid is scheduled onto. A chip typically contains many
+    AMD compute units or NVIDIA streaming multiprocessors; it's the final SKU.
+  }];
+
+  let mnemonic = "target_core";
+  let cppNamespace = "::mlir::iree_compiler::IREE::GPU";
+
+  let parameters = (ins
+    // Features
+    "ComputeBitwidthsAttr":$compute,
+    "StorageBitwidthsAttr":$storage,
+    "SubgroupOpsAttr":$subgroup,
+    "DotProductOpsAttr":$dot,
+    "MMAOpsArrayAttr":$mma,
+
+    // Limits
+    // Supported subgroup size choices.
+    "DenseI32ArrayAttr":$subgroup_size_choices,
+    // The maximal number of threads per X/Y/Z dimension in one workgroup.
+    "DenseI32ArrayAttr":$max_workgroup_sizes,
+    // The maximal number of threads we can have in one workgroup.
+    "uint32_t":$max_thread_size,
+    // The maximal number of shared memory bytes we can allocate per workgroup.
+    "uint32_t":$max_workgroup_memory_bytes
+  );
+
+  let assemblyFormat = "`<` struct(params) `>`";
+}
+
+//===----------------------------------------------------------------------===//
+// GPU Chip Level Feature/Limit Attributes
+//===----------------------------------------------------------------------===//
+
+// This section lists hardware features/limits at a single GPU chip level.
+// Here a GPU chip means the hardware functionality scope where the whole
+// software compute grid is scheduled onto. A chip typically contains many
+// AMD compute units or NVIDIA streaming multiprocessors; it's the final SKU.
+
+def IREEGPU_TargetChipAttr : AttrDef<IREEGPU_Dialect, "TargetChip"> {
+  let summary = "Chip level target description";
+  let description = [{
+    This attribute contains hardware features/limits at a single GPU core level.
+    Here a GPU core means the basic hardware functionality unit where a software
+    workgroup is scheduled onto; that is, a compute unit for AMD GPUs or a
+    streaming multiprocessor for NVIDIA GPUs.
+  }];
+
+  let mnemonic = "target_chip";
+  let cppNamespace = "::mlir::iree_compiler::IREE::GPU";
+
+  let parameters = (ins
+    "uint32_t":$core_count
+  );
+
+  let assemblyFormat = "`<` struct(params) `>`";
+}
+
+//===----------------------------------------------------------------------===//
+// GPU Target Attributes
+//===----------------------------------------------------------------------===//
+
+def IREEGPU_TAPI_CUDA    : I32EnumAttrCase<"CUDA",    0, "cuda">;
+def IREEGPU_TAPI_DirectX : I32EnumAttrCase<"DirectX", 1, "directx">;
+def IREEGPU_TAPI_HIP     : I32EnumAttrCase<"HIP",     2, "hip">;
+def IREEGPU_TAPI_Metal   : I32EnumAttrCase<"Metal",   3, "metal">;
+def IREEGPU_TAPI_Vulkan  : I32EnumAttrCase<"Vulkan",  4, "vulkan">;
+def IREEGPU_TAPI_WebGPU  : I32EnumAttrCase<"WebGPU",  5, "webgpu">;
+
+def IREEGPU_TargetAPI : I32EnumAttr<
```

Comment:
let's not use enums for things like this - that makes it impossible for out-of-tree/plugins to add targets, unless there's no plan for this to ever exist outside of in-tree code

Reply:
Dropped now.